### PR TITLE
Prevent duplicate document content within a partition

### PR DESCRIPTION
**Pull request by hedhoud**
## Summary

This PR prevents the same document content from being indexed more than once in a partition, even when the uploads use different filenames or file IDs.

Deduplication is intentionally scoped to a partition. The same document can still be indexed in another partition because partitions may represent separate users, workspaces, or retrieval contexts.

## Why this is needed

OpenRAG currently identifies documents by their file ID. Two uploads containing identical bytes can therefore enter the indexing pipeline under different IDs, consuming parser, embedding, storage, and queue capacity twice. A simple database uniqueness check is not sufficient because concurrent uploads can both start indexing before either document is stored.

This change addresses both completed duplicates and uploads that are still in progress.

## Behavior

OpenRAG calculates a SHA-256 digest while the upload is already being written to disk. This avoids a second full read and keeps memory usage constant for large documents.

Before queue submission, the content is reserved atomically for its partition. If another completed or in-progress document already has the same digest, the request is rejected with a conflict response that identifies the existing file. PostgreSQL transaction locking and a short-lived claim prevent concurrent requests from both entering expensive indexing work.

Claims are released after completion, failure, cancellation, or a submission error. Expired claims are also recoverable after an interrupted worker. Replacement remains valid when the content belongs to the same file ID, and copy and URL-indexing paths follow the same rule.

The digest is backend-owned. Metadata updates cannot inject or replace it.

## Configuration

Deduplication is enabled by default. It can be disabled with CONTENT_DEDUPLICATION_ENABLED=false, primarily for controlled benchmarks that need to exclude hashing overhead or intentionally index repeated content.

## Database migration

The migration adds a nullable content digest to the document catalog, a partition-scoped uniqueness constraint for populated digests, and a small table for active indexing claims.

Existing rows remain nullable. This makes the migration safe and fast for existing deployments, but documents indexed before this change will not be compared until a separate controlled backfill is performed. New uploads, replacements, copies, and URL-indexed documents are protected immediately.

## Compatibility

Document IDs, routes, partition behavior, and retrieval behavior remain unchanged. No frontend change is required. Duplicate uploads now return a conflict instead of starting another indexing task.

## Validation

- Full unit suite: 1,751 passed
- Final focused regression suite after concurrency hardening: 79 passed
- Ruff lint and format checks passed
- Python compilation passed
- Alembic reports a single migration head
- Migration was applied successfully to PostgreSQL 15
- Real PostgreSQL claim/release behavior was verified, including rejection of a concurrent duplicate and acceptance after claim release
- Database constraints were verified to reject the same digest in one partition while allowing it in another
- Diff integrity check passed

## Follow-up

A separate backfill should calculate hashes for existing catalog rows from their stored source files. Keeping that work separate avoids turning the release migration into a long-running filesystem scan.

## Summary by CodeRabbit

* **New Features**
  * Added SHA-256–based content deduplication for uploaded and copied documents.
  * Identical content within the same partition is deduplicated by default, returning HTTP 409 that references the existing file (including when using different file IDs).
  * Re-uploading identical content under the same file ID is treated as a valid replacement.

* **Bug Fixes**
  * Improved cleanup of temporary uploaded/downloaded files when indexing fails, is cancelled, or conflicts.
  * More predictable upload error handling (e.g., correct 409/413 behavior) and safer error responses.

* **Documentation**
  * Updated environment variable docs with `CONTENT_DEDUPLICATION_ENABLED` (default `true`).
  * `SAVE_UPLOADED_FILES` now defaults to `true`.

### CODE REVIEW

File: openrag/services/persistence/migrations/alembic/versions/c3d4e5f6a7b8_add_document_content_hashes.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        op.create_table(
+            "file_content_claims",
+            sa.Column("partition_name", sa.String(), nullable=False),
+            sa.Column("content_sha256", sa.String(length=64), nullable=False),
+            sa.Column("file_id", sa.String(), nullable=False),
+            sa.Column(
+                "expires_at",
+                sa.DateTime(timezone=True),
+                server_default=sa.text("now() + interval '24 hours'"),
+                nullable=False,
```
[the remark this answers is not in the document]
Comment (hedhoud) [reply openrag/services/persistence/migrations/alembic/versions/c3d4e5f6a7b8_add_document_content_hashes.py:44]: Fixed in 3235dff7. Active claims are now renewed by the shared pool for as long as the worker task is queued or running, so a long OCR backlog cannot expire the claim. The renewal stops when the task settles, and the behavior is covered by a regression test.

File: openrag/api/routers/admin/indexing.py [hunk 2]
```diff
@@ -139,7 +140,17 @@ async def add_file(
...
+            saved_upload = await save_file_to_disk_with_sha256(
+                file,
+                Path(config.paths.data_dir),
+                with_random_prefix=True,
+            )
+            file_path = saved_upload.path
+            content_sha256 = saved_upload.sha256
+        else:
+            file_path = await save_file_to_disk(file, Path(config.paths.data_dir), with_random_prefix=True)
+            content_sha256 = None
```
[the remark this answers is not in the document]
Comment (hedhoud) [reply openrag/api/routers/admin/indexing.py:172]: Fixed in 3235dff7. Workspace IDs are now parsed and validated before the upload is written. The regression test also verifies that an invalid workspace request creates no file on disk.

File: openrag/api/routers/admin/indexing.py [hunk 3]
```diff
@@ -173,16 +184,21 @@ async def add_file(
...
+            metadata=metadata,
+            sanitized_filename=file.filename,
+            original_filename=original_filename,
+            user=user,
+            workspace_ids=parsed_workspace_ids,
+            content_sha256=content_sha256,
+        )
+    except ConflictError:
+        file_path.unlink(missing_ok=True)
+        raise
```
[the remark this answers is not in the document]
Comment (hedhoud) [reply openrag/api/routers/admin/indexing.py:201]: Fixed in 3235dff7. POST, PUT, and MCP indexing now remove temporary files for any dispatch failure, including cancellation, while successful dispatch still transfers cleanup ownership to the worker. Regression tests cover the non-conflict and cancellation paths.

File: tests/unit/core/config/test_indexation.py [hunk 4]
```diff
@@ -29,3 +31,15 @@ def test_transcriber_config_drops_empty_components():
...
+
+
+def test_content_deduplication_is_enabled_by_default():
+    assert LoaderConfig().content_deduplication_enabled is True
+
+
+def test_content_deduplication_can_be_disabled_by_env(monkeypatch, tmp_path):
+    monkeypatch.setenv("CONTENT_DEDUPLICATION_ENABLED", "false")
+
+    settings = load_config(config_path=tmp_path)
```
[the remark this answers is not in the document]
Comment (hedhoud) [reply tests/unit/core/config/test_indexation.py:43]: I verified this against the public function imported by the test.  accepts ;  belongs to the internal loader and raises  through this public entry point. The current call is therefore correct, and the config test passes unchanged.

File: openrag/services/workers/dispatcher.py [hunk 5]
```diff
@@ -468,6 +538,16 @@ async def cancel_task(self, task_id: str) -> bool:
...
+        metadata = (details or {}).get("metadata") or {}
+        content_sha256 = metadata.get("content_sha256")
+        file_id = (details or {}).get("file_id")
+        partition = (details or {}).get("partition")
+        if content_sha256 and file_id and partition:
+            await self._document_repo.release_content_sha256_claim(
+                file_id=file_id,
+                partition=partition,
+                content_sha256=content_sha256,
+            )
```
[the remark this answers is not in the document]
Comment (hedhoud) [reply openrag/services/workers/dispatcher.py]: Fixed in 040250e4. Cancellation now only sends the Ray cancellation request. The shared pool keeps the claim until the worker reference actually settles, then releases it. This also covers tasks cancelled before the worker body starts, and a regression test verifies that the claim is not released early.

**Review by Ahmath-Gadji — APPROVED** [review]
Reviewed end-to-end across two passes (diff, all inline threads, and independent verification):

- Every issue raised by CodeRabbit and Codex (claim-expiry during long jobs, cancellation-vs-release ordering, upload cleanup on disconnect/cancel, workspace validation ordering, `put_file` error mapping, dispatch-failure cleanup) is fixed and confirmed in code.
- Independent pass: verified the content-hash lifecycle (claim/renew/release) across `document_repo.py`, `dispatcher.py`, and `indexer_pool.py`, the partition-scoped unique index + claims table migration (single true alembic head), file-delete/replace interaction with the unique constraint, and empty-file hashing edge case — no issues found.
- Ran the full local unit suite (1,853 tests) and `ruff check` clean on the reviewed commit.
- Branch was since rebased onto `develop` (new head `95a6fb0`) — diffed old vs new head and confirmed the rebase only picked up unrelated upstream commits (pymupdf font-digest fix, DI container changes); none of the dedup-related files changed.

Approving based on this review; CI on the rebased commit was still finishing at approval time.

**Comment by hedhoud** [comment]
@CodeRabbit review
